File: extension/background/tracking/graph/events/current-page.js
function applySetCurrentPageFallback(state, event) {
  const tabId = event.tabId;
  const targetNodeId = event.targetNode.nodeId;
  const previousNodeId = state.tabState[tabId]?.nodeId ?? null;
  const previousPageUrl = normalizePageUrlForIndex(state.tabState[tabId]?.url || "");
  const targetPageUrl = normalizePageUrlForIndex(event.url);
  const isNewNodeVisit = previousNodeId === null || previousNodeId !== targetNodeId;
  const isNewPageVisit = previousPageUrl !== targetPageUrl;

  upsertNodeFallback(state.graph, event.targetNode, event.occurredAt);

  if (isNewNodeVisit) {
    state.graph.nodes[targetNodeId].visitCount += 1;
  }

  if (isNewNodeVisit || isNewPageVisit) {
    state.graph.updatedAt = event.occurredAt;
  }

  state.tabState[tabId] = {
    nodeId: targetNodeId,
    url: event.url,
    updatedAt: event.occurredAt,
  };

  return state;
}

function upsertNodeFallback(graph, targetNode, occurredAt) {
  const normalizedLandingPageUrl =
    normalizePageUrlForIndex(targetNode.sampleUrl || "") || targetNode.sampleUrl || "";

  if (!graph.nodes[targetNode.nodeId]) {
    graph.nodes[targetNode.nodeId] = {
      nodeId: targetNode.nodeId,
      origin: targetNode.origin,
      host: targetNode.host,
      hostname: targetNode.hostname,
      protocol: targetNode.protocol,
      sampleUrl: targetNode.sampleUrl,
      defaultLandingPageUrl: normalizedLandingPageUrl,
      visitCount: 0,
      firstSeenAt: occurredAt,
      lastSeenAt: occurredAt,
    };

    return;
  }

  graph.nodes[targetNode.nodeId].lastSeenAt = occurredAt;
  graph.nodes[targetNode.nodeId].sampleUrl = targetNode.sampleUrl;
  if (!graph.nodes[targetNode.nodeId].defaultLandingPageUrl) {
    graph.nodes[targetNode.nodeId].defaultLandingPageUrl = normalizedLandingPageUrl;
  }
}

File: extension/wasm/visit-graph-engine/src/events/current_page.rs
use crate::{NodeSeed, TabStateEntry, TrackingState, db::upsert_node};

pub(crate) fn set_current_page(
    state: &mut TrackingState,
    tab_id: String,
    target_node: NodeSeed,
    occurred_at: String,
    url: String,
) {
    let previous_node_id = state
        .tab_state
        .get(&tab_id)
        .map(|entry| entry.node_id.clone());
    let previous_page_url = state.tab_state.get(&tab_id).map(|entry| entry.url.clone());
    let is_new_page_visit = previous_page_url.as_deref() != Some(url.as_str());

    upsert_node(&mut state.graph, &target_node, &occurred_at);

    let is_new_node_visit = previous_node_id
        .as_ref()
        .map(|node_id| node_id != &target_node.node_id)
        .unwrap_or(true);

    if is_new_node_visit {
        if let Some(node) = state.graph.nodes.get_mut(&target_node.node_id) {
            node.visit_count += 1;
        }
    }

    if is_new_node_visit || is_new_page_visit {
        state.graph.updated_at = Some(occurred_at.clone());
    }

    state.tab_state.insert(
        tab_id,
        TabStateEntry {
            node_id: target_node.node_id.clone(),
            url,
            updated_at: occurred_at,
        },
    );
}
